### ISSUE TITLE
Default format

### ISSUE THREAD

**Issue by Peque** [and3rson/flake8-colors#2]
It would be great if there was a default formatting, to avoid having to write the long, not-very-readable line in the configuration file:

```
format = ${cyan}%(path)s${reset}:${yellow_bold}%(row)d${reset}:${green_bold}%(col)d${reset}: ${red_bold}%(code)s${reset} %(text)
```

I do not care that much about the specific colors/bold. I guess it is reasonable to assume most people would be happy with any default colors (like the ones you chose).

- Maybe something like `format = colored`.
- Or it could be automatic by simply having `flake8-colors` installed.
- Or maybe have an option to run `flake8 --color` that automatically uses that line?

**Comment by Peque** [comment]
@sigmavirus2 would not make sense to integrate this in `flake8`? The current non-colored output is not awesome... :stuck_out_tongue_winking_eye:

**Comment by Peque** [comment]
@sigmavirus24 Would not make sense to integrate this in flake8? The current non-colored output is not awesome... :stuck_out_tongue_winking_eye:

**Comment by sigmavirus24** [comment]
I considered writing my own plugin before finding this. I don't think there's going to be a color scheme that won't aggravate someone. This plugin does the job well enough. Flake8's had plain-text output for years with few complaints. I don't think it's a necessary feature and I think this works very well as it is. @and3rson is welcome to move it to the [PyCQA](/pycqa) if they need help maintaining it though.

**Comment by Peque** [comment]
@sigmavirus24 I understand. Of course I was not thinking about setting the colored output as default. I was just thinking about integrating this in flake8 so that users could simply use `--color` (or alike) if they wanted to have some colors in the output. :blush:

If that is not going to happen, then I still want to propose @and3rson to allow the use of `format = default` or `format = colored` (or alike) to avoid having to write that long line if you do not care that much about the colors used.

**Comment by and3rson** [comment]
Thanks for your input! I'm glad people find my plugin useful.
On one hand, I like the way of automatically highlighting the output when the plugin is installed, but it brings some implicitness. I think that more obvious & pythonic way to achieve this would be  having something like @Peque mentioned with "default" or "{default}" placeholder.

**Comment by Peque** [comment]
@and3rson Thanks to you for sharing it! :blush:

Yeah, at least having a "default" would be great. On the other hand, about the implicitness of automatically highlighting the output when it is installed, I think other `flake8` plugins do exactly that: `pep8-naming`, `flake8-bugbear`, `flake8-import-order`. So could be better for consistency with other plugins? Perhaps it is harder to implement, though, in which case might not be worth it.

**Comment by myii** [comment]
@and3rson Firstly, thanks for this colourful plugin!

I was toying with opening a new issue but I think that my issue is related to the discussion above.

#### Prevents errors displaying in Vim (Syntastic)

Unfortunately, I can't use this plugin as-is because it effectively prevents any violations being displayed in the Vim editor when using [Syntastic](https://github.com/vim-syntastic/syntastic) syntax checker.  If I use the suggested default `format`, I get no errors at all.  To understand what was happening, I used a very simplified format instead:

```ini
format = ${cyan}%(path)s:%(row)d:%(col)d: %(code)s %(text)s
```

That allowed the errors to be displayed but with corruption:

```syntax
^[[0;36m/home/.../urls.py|1 col 1 warning| 1 blank line required between summary line and description [D205]
```

As a comparison, using `yellow` instead of `cyan`:

```syntax
^[[0;33m/home/.../urls.py|1 col 1 warning| 1 blank line required between summary line and description [D205]
```

#### Relationship to this discussion

> On one hand, I like the way of automatically highlighting the output when the plugin is installed, but it brings some implicitness.

* Implicitness is fine

> Yeah, at least having a "default" would be great.

* That's good, too

> * Or maybe have an option to run `flake8 --color` that automatically uses that line?

This suggestion from above gave me the idea of how this Syntastic problem could be solved.  Would it be possible to have something like a `--no-color` switch?  Then:

1. Running `flake8` from the terminal would display in all of its colourful glory
1. The `--no-color` option could be passed to Syntastic to allow that function as usual

```vim
let g:syntastic_python_flake8_args = '--no-color'
```

Perhaps there are other places where coloured output would cause problems.  Having the `--no-color` fallback would be useful in those situations, too.

**Comment by sigmavirus24** [comment]
In reality, this plugin should check for `stdin.istty()` (or whatever it's actually called) because on CI systems, it may break as well.

**Comment by sigmavirus24** [comment]
Also, regarding the "implicitness" of other plugins, they can also register checks that are off by default. It's not apparent how someone could have 6 formatters all of which say "I am the default" who wins in that situation?

**Comment by myii** [comment]
> In reality, this plugin should check for `stdout.istty()` (or whatever it's actually called) because on CI systems, it may break as well.

Thanks @sigmavirus24.  That's more useful than what I was proposing.  No need to configure Vim Syntastic, CI systems, etc.  Everything will _just work_.

#### Basic working solution

@and3rson The following changes are working for me:

```diff
@@ -1,3 +1,4 @@
+import sys
 from flake8.formatting import base
 from re import sub
 
@@ -33,7 +34,11 @@
 
     @classmethod
     def parse_options(cls, options):
-        options.format = sub(r'\$\{([\w]+)\}', cls._replace, options.format)
+        # Only use color formatting if invoked interactively
+        if sys.__stdin__.isatty():
+            options.format = sub(r'\$\{([\w]+)\}', cls._replace, options.format)
+        else:
+            options.format = 'default'
 
     @classmethod
     def _replace(cls, match):
```

#### Caveat: `isatty` may not be 100% reliable

https://stackoverflow.com/questions/3818511/how-to-tell-if-python-script-is-being-run-in-a-terminal-or-via-gui

> ...
> 
> A more reliable approach might therefore be to explicitly _tell_ the program whether it must output to `stdout` or where else, e.g. with a command-line flag.

#### Using `sys.__stdin__.isatty()`

https://stackoverflow.com/questions/6108330/checking-for-interactive-shell-in-a-python-script

> ```python
> import os, sys
> if os.isatty(sys.stdout.fileno()):
>     ...
> ```
> ...
> `sys.stdout.isatty()` is shorter.
> ...
> And `sys.__stdin__.isatty()` more reliable. (not affected by `stdin` redirect and allows piping out)

**Comment by sigmavirus24** [comment]
I would strongly advise against using the dunder variable here, but I would guess that `sys.stdout.isattty()` would be reliable enough.

**Comment by myii** [comment]
@sigmavirus24 Fair enough.  In terms of `stdout` vs. `stdin`, a large number of the resources I've found prefer the latter.  A few examples follow.

https://stackoverflow.com/questions/3818511/how-to-tell-if-python-script-is-being-run-in-a-terminal-or-via-gui:

> I scoured SE for an answer to this but everywhere indicated the use of `sys.stdout.isatty()` or `os.isatty(sys.stdout.fileno())`. Neither of these dependably caught my GUI test cases.
> 
> Testing standard input was the only thing that worked for me:
> ```python
> sys.stdin.isatty()
> ```

https://stackoverflow.com/questions/699390/whats-the-best-way-to-tell-if-a-python-program-has-anything-to-read-from-stdin

> If you want to detect if someone is piping data into your program, or running it interactively you can use isatty to see if stdin is a terminal:
> 
> ```bash
> $ python -c 'import sys; print sys.stdin.isatty()'
> True
> $ echo | python -c 'import sys; print sys.stdin.isatty()'
> False
> ```

Also, amongst others:

* https://stackoverflow.com/questions/1285024/how-can-i-check-to-see-if-a-python-script-was-started-interactively
* http://pleac.sourceforge.net/pleac_python/userinterfaces.html

**Comment by sigmavirus24** [comment]
The first stackoverflow link is leagues more relevant than the latter (flake8-colors shouldn't care if there's anything to read from stdin). It would also be a significantly more useful answer if they explained their testing methodology and what GUIs they tested with.

**Comment by myii** [comment]
For what it's worth, I've tested each option (including the maligned dunder) and all are working for my needs.  I'm happy to defer this decision to the Flake8 experts, of which I am not one!
